README.md
# Tvheadend viewer (Tvheadendv)

Copyright (c) 2012 Miguel Angel Gabriel (magabriel@gmail.com). All rights reserved.

Released under the GNU GPL V3. 

## Synopsis

This is a mobile web viewer for the excelent HTS Tvheadend {https://www.lonelycoder.com/tvheadend/} streaming server.

Tvheadend does not provide an adequate mobile version and there are no plans to implement one. There are both an official iPhone client and at least another unofficial one for Android {http://john-tornblom.github.com/TVHGuide}. 

I started this project as an excuse to teach myself jQuery Mobile. I would be happy if anyone finds it useful.

Screenshots are available in /doc/screenshots folder.

## Features

- Easy integration into TvHeadend installation.
- Show channel tags, channels and EPG.
- When in the channel or EPG list, tapping on an item will open the default video player to play the live stream directly from your server (I personally recommend the excellent MX Videoplayer in the Android platform). 

## Requirements

- A working installation of TvHeadend. 
- A mobile platform with a modern browser (tested under Androd 2.3 / 4.0 and stock browser / Dolphin browser).
- An external media player that can play streaming urls. 

## Installation

- Instructions are for Ubuntu. Not tested in other distributions. This explanation assumes that TvHeadend is installed in `/usr/share/tvheadend`, so adjust it according to your installation.
- Download everything under the repository's "public_html" folder and copy somewhere in the same machine that runs Tvheadend (say, "/home/me/apps/tvheadendv"). Do not forget to chmod 664 everything in there.
- From here you have two options (choose either one):

    1.- Make a symbolic link pointing TvHeadend to the directory containing the application:
    
            sudo ln -s /home/me/apps/tvheadendv/public_html /usr/share/tvheadend/src/webui/static/tvhv

    2.- Copy the contents of the public_html folder just downloaded to the right directory inside TvHeadend installation:
    
            sudo cp -r /home/me/apps/tvheadendv/public_html /usr/share/tvheadend/src/webui/static/tvhv
                
  Either way, after that you will be able to access the application opening {http://yourserver:9981/static/tvhv/index.html} from your mobile browser.
       
 

 




  



CHANGELOG.md
# CHANGELOG

V1.1 (2012-11-28)

- Huge refactorings, improvements and performance optimizations everywhere.
- Got rid of the PHP JSONP proxy. Pure javascript solution now (yes!!).
- Ready for easy integration into TvHeadend.
- Upgraded to JQuery 1.8.3 and JQuery Mobile 1.2.0.

V1.0 (2012-04-14)
 
- First truly functional version.
- Upgraded to JQuery 1.7.1 and JQuery Mobile 1.1.0 RC 2.
- I18N working: to support another language just add a new translation file.
- A whole lot of wonderful features added and nasty bugs fixed.  

public_html/js/utils.js
/*
 * ----------------------------------------------------------------------------
 * Copyright (c) 2012 Miguel Angel Gabriel. All Rights Reserved. 
 * Released under the GNU GPL V3. 
 * ----------------------------------------------------------------------------
 */
var utils = {
		
	getISODateTime: function(d)
	{
	    // padding function
	    var s = function(a,b){return(1e15+a+"").slice(-b);};

	    // default date parameter
	    if (typeof d === 'undefined'){
	        d = new Date();
	    };

	    // return ISO datetime
	    return d.getFullYear() + '-' +
	        s(d.getMonth()+1,2) + '-' +
	        s(d.getDate(),2) + ' ' +
	        s(d.getHours(),2) + ':' +
	        s(d.getMinutes(),2) + ':' +
	        s(d.getSeconds(),2);
	},
	
	getDisplayTime: function(d)
	{
	    // padding function
	    var s = function(a,b){return(1e15+a+"").slice(-b);};

	    // default date parameter
	    if (typeof d === 'undefined'){
	        d = new Date();
	    } else if (typeof d === 'number'){
	    	d = new Date(d * 1000);
	    } 

	    // return ISO datetime
	    return s(d.getHours(),2) + ':' + s(d.getMinutes(),2);
	},
	
	getDisplayDuration: function(duration)
	{
		h = Math.floor(duration / 3600);
		s1 = duration - (h*3600);
		m = Math.floor(s1 / 60);
		
		h = (h < 10) ? '0'+h : h;
		m = (m < 10) ? '0'+m : m;
		
		return h+':'+m;
	},
	
	getDisplayRemaining: function(start, end)
	{
		dStart = new Date(start * 1000);
		dEnd = new Date(end * 1000);
		dNow = new Date();
		duration = dEnd - dStart;
		
		remaining = duration - (dNow - dStart);
		
		return this.getDisplayDuration(remaining / 1000);
	},
	
	getElapsedDuration: function(start, end, scale)
	{
		dStart = new Date(start * 1000);
		dEnd = new Date(end * 1000);
		dNow = new Date();
		duration = dEnd - dStart;
		
		if (dStart > dNow) {
			return 0;
		}
		
		if (dEnd < dNow) {
			return scale;
		}
		
		elapsed = (duration > 0) ? ((dNow - dStart) / duration) : 0; 
		
		result = Math.floor(scale * elapsed);
		
		return result;
	},
	
	sortByKey : function (array, key, zeroLast) {
	    return array.sort(function(a, b) {

	    	var keys = key.split(',');
	    	
	    	var x='';
	    	var y='';
	    	
	    	var padLength = 20;
	    	$.each(keys, function(i, key) {		
	    		
	    		var keyParts = key.split(':');
	    		var keyName = keyParts[0];
	    		var keyType = 's';
	    		if (keyParts.length > 0) {
	    			keyType = keyParts[1];
	    		}
	    		
	    		var va = a[keyName].toString();
	    		var vb = b[keyName].toString();

	    		if (zeroLast) {
	    			if (va == 0) {
	    				va = utils.strRepeat('z', padLength); 
	    			}
	    			if (vb == 0) {
	    				vb = utils.strRepeat('z', padLength); 
	    			}
	    		};
	    		
	    		padA = utils.strRepeat(' ', padLength - va.length);
	    		padB = utils.strRepeat(' ', padLength - vb.length);
	    		
	    		x+= (keyType == 'n') ? padA + va : va + padA ;
	    		y+= (keyType == 'n') ? padB + vb : vb + padB ;
	    	});
	    	
	    	x = x.toUpperCase();
	    	y = y.toUpperCase();
	    		         
	        return ((x < y) ? -1 : ((x > y) ? 1 : 0));
	    });
	},
	
	strRepeat : function(str, num )	{
	    return new Array( isNaN(num) ? 1: num + 1 ).join( str );
	},
	
	isArray: function(input){
		return typeof(input)=='object'&&(input instanceof Array);
	},
	
	runTasks: function ( tasks, idx ) {
	   idx = idx || 0;
	   task = tasks[idx]; 
	   if (utils.isArray(task)) {
		   task[0](task[1]);
	   } else {
		   task();
	   }
	   idx++;
	   me = this;
	   if( idx < tasks.length ) {
	      setTimeout( function(){ me.runTasks(tasks, idx); },1);
	   }
	},
	
	getCurrentBaseUrl: function () {
		location = window.location;
		port = location.port ? ':'+location.port : '';
		ret = location.protocol + '//' + location.hostname + port + location.pathname;
		return ret;
	},
	
	isNumber : function(n) {
		return !isNaN(parseFloat(n)) && isFinite(n);
	}

};

/*
 * A plugin to manage localization with jquery.localize.js
 */
(function($) {

	var options = { pathPrefix : "js", 		// Where the transaltion files are
					skipLanguage: /^en/}; 	// Default language (to avoid trying to translate it)
	
	/*
	 * jquery.localize doesn't work on buttons that have already been enhanced by
	 * JQuery Mobile. We need to override the translation callback to provide an
	 * additional case for the button's markup. 
	 */
	options.callback = function(data, defaultCallback) {
			
		return $items.each(function() {
			var elem, key, value;
			elem = $(this);
			key = elem.attr("rel").match(/localize\[(.*?)\]/)[1];
			value = valueForKey(key, data);
			if (elem.is('input')) {
				if (elem.is("[placeholder]")) {
					return elem.attr("placeholder", value);
				} else {
					return elem.val(value);
				}
			} else if (elem.is('optgroup')) {
				return elem.attr("label", value);
			} else if (elem.is('img')) {
				value = valueForKey("" + key + ".alt", data);
				if (value != null) {
					elem.attr("alt", value);
				}
				value = valueForKey("" + key + ".src", data);
				if (value != null) {
					return elem.attr("src", value);
				}
			} else if (elem.is('a')) {
				/* If this is an already-enhanced button, modify just
				 * the html part of the inner span
				 */ 
				button = elem.find(".ui-btn-text");
				if (button && button.html()) {
					value = valueForKey(key, data);
					return button.html(value);
				} else {
					return elem.html(value);
				}
			} else {
				return elem.html(value);
			}
		});
	};
	
	/*
	 * This is just a copy of the same function inside jquery.localize, because
	 * it cannot be called directly from our callback.
	 */
    valueForKey = function(key, data) {
        var keys, value, _i, _len;
        keys = key.split(/\./);
        value = data;
        for (_i = 0, _len = keys.length; _i < _len; _i++) {
          key = keys[_i];
          value = value != null ? value[key] : null;
        }
        return value;
      };
	
    /*
     * The JQuery plugin call
     */
	$.fn.translate = function() {
		$items = this.find("[rel*=localize]");
		if ($items.length == 0) {
			$items = this;
		}
		return $items.localize("lang", options);
	};
	
	/*
	 * Used to force a target language, in case the autodetection
	 * cannot be reliable enough. 
	 */
	$.setLanguage = function(lang) {
		options.language = lang;
	};
	
})(jQuery);

/*
 * Several mini plugins
 */ 
(function($){
	$.fn.reverse = [].reverse;
	$.fn.sort = [].sort;
	$.fn.shift = [].shift;
})(jQuery);

public_html/jsonp-proxy.php
<?php
/**
 * ----------------------------------------------------------------------------
 * (C) 2012 Miguel Angel Gabriel. Released under the GNU GPL V3 
 * ----------------------------------------------------------------------------
 *
 * Class to call a JSON server using JSONP protocol when the real
 * server doesn't provide JSONP by itself.
 *
 * If you need to call a server that provides a JSON response but
 * you are doing it from a Javascript program that was not downloaded
 * from the same domain, the call will fail because of the "same origin"
 * security restriction (NB: "same origin" means same protocol, hostname
 * and port).
 *
 * To overcome this restriction you have two options:
 * 1) Use a JSONP call instead.
 * 2) Use a normal JSON call from another server, because the "same origin"
 *    restriction does not apply to servers.
 *
 * Note that 1) is only possible if the JSON server also supports JSONP or
 * if you are able to modify it to build JSONP supoort into it. If you cannot
 * include native JSONP support, this class is the solution.
 *
 * This class provides a proxy that acts as a JSONP server that will forward
 * a normal JSON call to the real server and return the JSON response
 * formatted as JSONP. To accomplish this:
 * - Copy this file to whatever PHP server you have access to (could be localhost)
 *   making it available via http (i.e. http://localhost/jsonp-proxy.php)
 * - Make a JSONP request to the proxy with the right parameters (see below)
 *   and it will make the JSON request to the real server an return a
 *   valid JSONP response.
 *
 */

$jsonp = new JsonpProxy();
return $jsonp->process();

class JsonpProxy
{
	protected $serverUrl;

	/**
	 * Process a JSONP request calling the real JSON server.
	 * 
	 * Expected url query parameters:
	 * - "jsoncallback" : The name to assign to the JSONP callback
	 * - "callback" : Same as jsoncallback (either one is valid)
	 * - "url" : The url for the real JSON server (urlencoded)
	 * - "api" : The api to call the real JSON server
	 * - any other : JSON request parameters
	 * 
	 * Example:
	 * - Request (broken down into several lines for readability):
	 * 
	 * 		http://<myhost>/?
	 * 				callback=mycallback&
	 * 				url=http%3A%2F%2Fjsonserverhost%3A9123%2F&
	 * 				api=items&
	 * 				from=100&
	 * 				limit=20
	 * 
	 * 	- This will call 'http://jsonserverhost:9123/items?from=100&limit=20'
	 *    and return the following JSONP response:
	 *    
	 *    	mycallback=({ ... JSON response ...})
	 * 		
	 */
	public function process()
	{
		$params = $_REQUEST;
		
		// Look if we have a JSONP request
		$jsoncallback = '';
		if (isset($params['jsoncallback'])) {
			$jsoncallback = $params['jsoncallback'];
			unset($params['jsoncallback']);
		} elseif (isset($params['callback'])) {
			$jsoncallback = $params['callback'];
			unset($params['callback']);
		} else {
			echo "ERROR: Missing jsoncallback parameter\n";
			header("HTTP/1.0 400 Bad request");
			return;
		}
		
		// Get the real server url
		if (isset($params['url'])) {
			$this->serverUrl = $params['url'];
			unset($params['url']);
		} else {
			echo "ERROR: Missing url parameter\n";
			header("HTTP/1.0 400 Bad request");
			return;
		}
		
		$requestUrl = $this->constructRequestUrl($params);
		
		$response = $this->sendRequest($requestUrl);
		
		/* Add the user preferred language (set in the browser but this is the only
		 * way to get to it. It can then be used in the client code to localize the 
		 * user interface.
		 * NOTE: I am not using json_ functions because they only work with UTF-8
		 * data. This solution is ugly but it works.
		 */
		if (strlen($response) > 0 && substr($response, 0, 1) == '{') {
			$response = '{ "acceptLanguage":'.'"'.$_SERVER['HTTP_ACCEPT_LANGUAGE'].'"'.
			','.substr($response, 1);
		}
		
		// Build the response
		$return = $jsoncallback.'('.$response.')';
		
		header("Content-Type: application/javascript");
		
		echo $return;
	}
	
	function prepareJSON($input) {
	
		$input = mb_convert_encoding($input, 'UTF-8', 'UTF-8');
	
		return $input;
	}
	
	function my_json_encode($arr)
	{
		//convmap since 0x80 char codes so it takes all multibyte codes (above ASCII 127). So such characters are being "hidden" from normal json_encoding
		array_walk_recursive($arr, function (&$item, $key) {
			if (is_string($item)) $item = mb_encode_numericentity($item, array (0x80, 0xffff, 0, 0xffff), 'UTF-8');
		});
		return mb_decode_numericentity(json_encode($arr), array (0x80, 0xffff, 0, 0xffff), 'UTF-8');
	
	}
	protected function sendRequest($url)
	{
		// Request
		$response = file_get_contents($url);
		$responseHeaders = $this->getResponseHeaders($http_response_header);
		
		if ( ! isset($responseHeaders['Content-Type']) || 
			(strpos($responseHeaders['Content-Type'], 'text/x-json') === false) ) {
		
			echo "ERROR: The server did not return a JSON response\n";
			print_r($responseHeaders);
			header("HTTP/1.0 400 Bad request");
			return;
		}
		
		return $response;
	}
	
	protected function constructRequestUrl($params)
	{
		// API to call
		$api = '';
		if (isset($params['api'])) {
			$api = $params['api'];
			unset($params['api']);
		}
		
		// Remaining parameters
		$reqParams = array();
		foreach ($params as $key=>$value)
		{
			$reqParams[]= $key.'='.$value;
		}
		
		// Build request
		$query = implode('&', $reqParams);
		$url = $this->serverUrl.'/'.$api.'?'.$query;
		
		return $url;
	}
	
	protected function getResponseHeaders($headers)
	{
		$arh = array();
		if (is_array($headers)) {
			foreach ($headers as $header) {
				$header = explode(":", $header);
				$arh[array_shift($header)] = trim(implode(":", $header));
			}
		}
		return $arh;
	}
}




public_html/js/tools.js
/*
 * ----------------------------------------------------------------------------
 * Copyright (c) 2012 Miguel Angel Gabriel. All Rights Reserved. 
 * Released under the GNU GPL V3. 
 * ----------------------------------------------------------------------------
 */
"use strict";

var Mags = Mags || {};
Mags.Tools = Mags.Tools || {};

(function($, self) {

	self.getISODateTime =
			function(d) {
				// padding function
				var s = function(a, b) {
					return (1e15 + a + "").slice(-b);
				};

				// default date parameter
				if (typeof d === 'undefined') {
					d = new Date();
				}
				;

				// return ISO datetime
				return d.getFullYear()
						+ '-'
						+ s(d.getMonth() + 1, 2)
						+ '-'
						+ s(d.getDate(), 2)
						+ ' '
						+ s(d.getHours(), 2)
						+ ':'
						+ s(d.getMinutes(), 2)
						+ ':'
						+ s(d.getSeconds(), 2);
			};

	self.getDisplayTime = function(d) {
		// padding function
		var s = function(a, b) {
			return (1e15 + a + "").slice(-b);
		};

		// default date parameter
		if (typeof d === 'undefined') {
			d = new Date();
		} else if (typeof d === 'number') {
			d = new Date(d * 1000);
		}

		// return ISO datetime
		return s(d.getHours(), 2) + ':' + s(d.getMinutes(), 2);
	};

	self.getDisplayDuration = function(duration) {
		var h = Math.floor(duration / 3600);
		var s1 = duration - (h * 3600);
		var m = Math.floor(s1 / 60);

		h = (h < 10) ? '0' + h : h;
		m = (m < 10) ? '0' + m : m;

		return h + ':' + m;
	};

	self.getDisplayRemaining = function(start, end) {
		var dStart = new Date(start * 1000);
		var dEnd = new Date(end * 1000);
		var dNow = new Date();
		var duration = dEnd - dStart;

		var remaining = duration - (dNow - dStart);

		return this.getDisplayDuration(remaining / 1000);
	};

	self.getElapsedDuration = function(start, end, scale) {
		var dStart = new Date(start * 1000);
		var dEnd = new Date(end * 1000);
		var dNow = new Date();
		var duration = dEnd - dStart;

		if (dStart > dNow) {
			return 0;
		}

		if (dEnd < dNow) {
			return scale;
		}

		var elapsed = (duration > 0) ? ((dNow - dStart) / duration) : 0;

		var result = Math.floor(scale * elapsed);

		return result;
	};

	self.sortByKey = function(array, key, zeroLast) {
		return array.sort(function(a, b) {

			var keys = key.split(',');

			var x = '';
			var y = '';

			var padLength = 20;
			$.each(keys, function(i, key) {

				var keyParts = key.split(':');
				var keyName = keyParts[0];
				var keyType = 's';
				if (keyParts.length > 0) {
					keyType = keyParts[1];
				}

				var va = a[keyName].toString();
				var vb = b[keyName].toString();

				if (zeroLast) {
					if (va == 0) {
						va = self.strRepeat('z', padLength);
					}
					if (vb == 0) {
						vb = self.strRepeat('z', padLength);
					}
				}
				;

				var padA = self.strRepeat(' ', padLength - va.length);
				var padB = self.strRepeat(' ', padLength - vb.length);

				x += (keyType == 'n') ? padA + va : va + padA;
				y += (keyType == 'n') ? padB + vb : vb + padB;
			});

			x = x.toUpperCase();
			y = y.toUpperCase();

			return ((x < y) ? -1 : ((x > y) ? 1 : 0));
		});
	};

	self.strRepeat = function(str, num) {
		return new Array(isNaN(num) ? 1 : num + 1).join(str);
	};

	self.isArray = function(input) {
		return typeof (input) == 'object' && (input instanceof Array);
	};

	self.runTasks = function(tasks, idx) {
		idx = idx || 0;
		var task = tasks[idx];
		if (self.isArray(task)) {
			task[0](task[1]);
		} else {
			task();
		}
		idx++;
		var me = this;
		if (idx < tasks.length) {
			setTimeout(function() {
				me.runTasks(tasks, idx);
			}, 1);
		}
	};

	self.getCurrentBaseUrl = function() {
		var location = window.location;
		var port = location.port ? ':' + location.port : '';
		var ret = location.protocol + '//' + location.hostname + port + location.pathname;
		return ret;
	};

	self.isNumber = function(n) {
		return !isNaN(parseFloat(n)) && isFinite(n);
	};

})(jQuery, Mags.Tools);

/*
 * Class to manage asynchronous fill of a listview
 */
(function($, self) {

	/**
	 * Constructor
	 * 
	 * @param $list
	 * @returns {self.ListFiller}
	 */
	self.ListFiller = function($list) {
		this.$list = $list;
		this.tasks = [];
		this.items = [];
	};

	self.ListFiller.prototype = {

		/**
		 * Add one item to the task queue. A task item is made of a data element
		 * with the data to be added to the listview and a function object that
		 * transform the data element into a text representation of the listview
		 * item to be added.
		 * 
		 * @param entry
		 * @param fnFillOne
		 */
		addItem : function(entry, fnFillOne) {
			this.tasks.push([ fnFillOne, entry ]);
		},

		/**
		 * Run the tasks in the queue sequentially
		 * 
		 * @param fnWrapUp
		 *            A function object to finish the list
		 */
		run : function(fnWrapUp) {
			// Push a task to append all the items to the listview and refresh it
			this.tasks.push([ function(me) {
				me.$list.empty().append(me.items.join('')).listview('refresh')
			}, this ]);
			// Push the real wrapup function
			this.tasks.push(fnWrapUp);
			// And run everything
			this._doRun(0);
		},

		/**
		 * Internal recursive function to execute all task in sequence
		 * 
		 * @param idx
		 *            The next task index
		 */
		_doRun : function(idx) {
			// Get the task and execute it
			var task = this.tasks[idx];
			if (typeof (task) == 'object' && (task instanceof Array)) {
				// A task with data
				this.items.push(task[0](task[1]));
			} else {
				// Without data
				this.items.push(task());
			}
			idx++;
			var me = this;
			if (idx < this.tasks.length) {
				// Call self to execute next task
				setTimeout(function() {
					me._doRun(idx);
				}, 1);
			}
		}
	};

})(jQuery, Mags.Tools);

/*
 * A plugin to manage localization with jquery.localize.js
 */
(function($) {

	var options = {
		pathPrefix : "js", // Where the translation files are
		skipLanguage : /^en/
	}; // Default language (to avoid trying to translate it)

	var $items;

	/*
	 * jquery.localize doesn't work on buttons that have already been enhanced by
	 * JQuery Mobile. We need to override the translation callback to provide an
	 * additional case for the button's markup. 
	 */
	options.callback = function(data, defaultCallback) {

		return $items.each(function() {
			var elem, key, value;
			elem = $(this);
			key = elem.attr("rel").match(/localize\[(.*?)\]/)[1];
			value = valueForKey(key, data);
			if (elem.is('input')) {
				if (elem.is("[placeholder]")) {
					return elem.attr("placeholder", value);
				} else {
					return elem.val(value);
				}
			} else if (elem.is('optgroup')) {
				return elem.attr("label", value);
			} else if (elem.is('img')) {
				value = valueForKey("" + key + ".alt", data);
				if (value != null) {
					elem.attr("alt", value);
				}
				value = valueForKey("" + key + ".src", data);
				if (value != null) {
					return elem.attr("src", value);
				}
			} else if (elem.is('a')) {
				/* If this is an already-enhanced button, modify just
				 * the html part of the inner span
				 */
				var button = elem.find(".ui-btn-text");
				if (button && button.html()) {
					value = valueForKey(key, data);
					return button.html(value);
				} else {
					return elem.html(value);
				}
			} else {
				return elem.html(value);
			}
		});
	};

	/*
	 * This is just a copy of the same function inside jquery.localize, because
	 * it cannot be called directly from our callback.
	 */
	var valueForKey = function(key, data) {
		var keys, value, _i, _len;
		keys = key.split(/\./);
		value = data;
		for (_i = 0, _len = keys.length; _i < _len; _i++) {
			key = keys[_i];
			value = value != null ? value[key] : null;
		}
		return value;
	};

	/*
	 * The JQuery plugin call
	 */
	$.fn.translate = function() {
		$items = this.find("[rel*=localize]");
		if ($items.length == 0) {
			$items = this;
		}
		return $items.localize("lang", options);
	};

	/*
	 * Used to force a target language, in case the autodetection
	 * cannot be reliable enough. 
	 */
	$.setLanguage = function(lang) {
		options.language = lang;
	};

})(jQuery);

/*
 * Several mini plugins
 */
(function($) {
	$.fn.reverse = [].reverse;
	$.fn.sort = [].sort;
	$.fn.shift = [].shift;
})(jQuery);